sound/soc/intel/boards/sof_sdw_cs_amp.c

```diff
@@ -0,0 +1,48 @@
+// SPDX-License-Identifier: GPL-2.0-only
+// Copyright (c) 2023 Intel Corporation
+
+/*
+ *  sof_sdw_cs_amp - Helpers to handle CS35L56 from generic machine driver
+ */
+
+#include <linux/device.h>
+#include <linux/errno.h>
+#include <sound/soc.h>
+#include <sound/soc-acpi.h>
+#include "sof_sdw_common.h"
+
+#define CODEC_NAME_SIZE	8
+
+static int cs_spk_init(struct snd_soc_pcm_runtime *rtd)
+{
+	const char *dai_name = rtd->dai_link->codecs->dai_name;
+	struct snd_soc_card *card = rtd->card;
+	char codec_name[CODEC_NAME_SIZE];
+
+
+	snprintf(codec_name, CODEC_NAME_SIZE, "%s", dai_name);
+	card->components = devm_kasprintf(card->dev, GFP_KERNEL,
+					  "%s spk:%s",
+					  card->components, codec_name);
+	if (!card->components)
+		return -ENOMEM;
+
+	return 0;
+}
+
+
+int sof_sdw_cs_amp_init(struct snd_soc_card *card,
+			const struct snd_soc_acpi_link_adr *link,
+			struct snd_soc_dai_link *dai_links,
+			struct sof_sdw_codec_info *info,
+			bool playback)
+{
+	/* Count amp number and do init on playback link only. */
+	if (!playback)
+		return 0;
+
+	info->amp_num++;
+	dai_links->init = cs_spk_init;
```

Comment:
humm, that doesn't sound good to me, we are going to have multiple 'spk' strings in the 'card->components'

this is a generic problem, for maxim we limit the use to 2 amps with this code:

````c
	info->amp_num++;
	if (info->amp_num == 2)
		dai_links->init = spk_init;
````

not sure how to solve this.

Reply:
> humm, that doesn't sound good to me, we are going to have multiple 'spk' strings in the 'card->components'

cs_spk_init is the .init callback of a dai_link. So, it will be called per dai_link. `dai_links->init = cs_spk_init;` is just to overwrite the init callback to the same dai_link because all cs35l56 amps belongs to the same dai link.

> 
> this is a generic problem, for maxim we limit the use to 2 amps with this code:
> 
> ```c
> 	info->amp_num++;
> 	if (info->amp_num == 2)
> 		dai_links->init = spk_init;
> ```
> 
> not sure how to solve this.

Reply:
humm, ok @bardliao i can understand the cirrus logic override, but I don't get why the maxim stuff only assigns the dailink init for the second amp. that seem to prevent a configuration with a single amp?

Reply:
> humm, ok @bardliao i can understand the cirrus logic override, but I don't get why the maxim stuff only assigns the dailink init for the second amp. that seem to prevent a configuration with a single amp?

@plbossart Yes, I think the driver assume the amp number is always 2. It always adds `max_98373_dapm_routes` to the dapm routes which contains "Left BE_OUT" and "Right BE_OUT".